chocolatey/choco
# (#158) Restore --source for choco list

## Description Of Changes

- Restore `--source` param for `choco list`
- Add a handful of interfaces for the functionally different pieces of `ISourceRunner`
- Add appropriate interfaces for each currently available ISourceRunner
- Add pre-checks into ChocolateyPackageService which verifies which methods that the source runner it's trying to invoke supports and throws an error (or warning, in the case of `DryRun` / `--noop` operations) before ever trying to run it if the source runner doesn't support that operation.
- Update a few affected mocks in tests to ensure they still work correctly.

## Motivation and Context
Since alternate sources can list installed packages separately to our normal package list, we need to allow list to specify --source. This will have no effect for normal sources, but will select alternate source runners if specified.

Additionally, the early checks for each source runner as to which operations they support was implemented in a way that allows us to gut ISourceRunner in a later version and fall back to having the supported methods for each source runner enforced by the type system, so we have a clear indication as to which source runners support each kind of operation via an appropriate interface cast. I haven't added obsoletions to ISourceRunner's main methods just yet, but that might be something we want to do in future.

## Testing

- Check that `choco list --source windowsfeatures` works
- Check that `choco search --source windowsfeatures` works
- Check that `choco list --source cygwin` reports as not supported _without_ first installing cygwin if it's not yet installed
- Ensure that supported operations for list and for search for each source work as expected (currently afaik only the windowsfeatures source does any actual searching, the rest only do listing for local or remote packages).

### Operating Systems Testing

Win11

## Change Types Made



## Change Checklist


## Related Issue

Restoring some functionality that was impacted by our changes when resolving #158
